src/umpire/util/AllocationMap.cpp

```diff
@@ -14,249 +14,481 @@
-    auto addr = record.key;
-    auto vec = *record.value;
+  uintptr_t key = 0;
+  for(m_last = judy_strt(m_array, reinterpret_cast<unsigned char*>(&key), 0);
+      m_last != nullptr;
+      m_last = judy_nxt(m_array)) {
+    auto list = reinterpret_cast<RecordList*>(*m_last);
+
+    void* addr;
+    judy_key(m_array, reinterpret_cast<unsigned char*>(&addr), m_depth * JUDY_key_size);
 
     std::stringstream ss;
-    ss << reinterpret_cast<void*>(addr) << " : {" << std::endl;
     bool any_match = false;
-    for (auto const& records : vec) {
-      AllocationRecord* tmp = reinterpret_cast<AllocationRecord*>(records);
-      if (pred(tmp)) {
+    ss << addr << " {" << std::endl;
+    for (auto iter{list->begin()}; iter != list->end(); ++iter) {
+      if (pred(*iter)) {
         any_match = true;
-        ss << "  " << tmp->m_size <<
-          " [ " << reinterpret_cast<void*>(addr) <<
-          " -- " << reinterpret_cast<void*>(addr+tmp->m_size) <<
+        ss << iter->m_size <<
+          " [ " << reinterpret_cast<void*>(iter->m_ptr) <<
+          " -- " << reinterpret_cast<void*>(static_cast<unsigned char*>(iter->m_ptr)+iter->m_size) <<
           " ] " << std::endl;
       }
     }
     ss << "}" << std::endl;
-    if (any_match) { os << ss.str(); }
+
+    if (any_match) {
+      os << ss.str();
+    }
   }
 }
 
-void
-AllocationMap::printAll(std::ostream& os) const
+void AllocationMap::printAll(std::ostream& os) const
 {
   os << "🔍 Printing allocation map contents..." << std::endl;
 
-  print([] (const AllocationRecord*) { return true; }, os);
+  print([] (const AllocationRecord&) { return true; }, os);
 
   os << "done." << std::endl;
 }
 
+
+AllocationMapConstIterator AllocationMap::begin() const
+{
+  return AllocationMapConstIterator{this, false};
+}
+
+AllocationMapConstIterator AllocationMap::end() const
+{
+  return AllocationMapConstIterator{this, true};
+}
+
+// AllocationMapConstIterator
+AllocationMapConstIterator::AllocationMapConstIterator(const AllocationMap* map, bool end)
+  : m_array(map->m_array), m_last(nullptr), m_ptr(0), m_iter(nullptr)
+{
+  if (!end) {
+    m_last = judy_strt(m_array, reinterpret_cast<const unsigned char*>(&m_ptr), 0);
+  }
+  else {
+    m_last = judy_end(m_array);
+  }
+
+  if (m_last) {
+    m_list = reinterpret_cast<RecordList*>(*m_last);
+    m_iter = new RecordListConstIterator{m_list, end};
+  }
+}
+
+AllocationMapConstIterator::~AllocationMapConstIterator()
+{
+  delete m_iter;
+}
+
+AllocationRecord& AllocationMapConstIterator::operator*() const
+{
+  return m_iter->operator*();
+}
+
+const AllocationRecord* AllocationMapConstIterator::operator->() const
+{
+  return m_iter->operator->();
+}
+
+AllocationMapConstIterator& AllocationMapConstIterator::operator++()
+{
+  (*m_iter)++;
+  if (*m_iter == m_list->end()) {
+    // Move to a new pointer
+    JudySlot* new_slot = judy_strt(m_array, reinterpret_cast<const unsigned char*>(&m_ptr), 0);
```

Comment:
Is this the correct approach to iterate over the whole array?

Reply:
Information on judy indicates this is not always the case. It only iterates over level 0 trees in the array. Fixed in a new version